src/compiler/checker.ts

```diff
@@ -2428,10 +2428,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return visitEachChild(node, markAsSynthetic, /*context*/ undefined);
     }
 
-    function getEmitResolver(sourceFile: SourceFile, cancellationToken: CancellationToken) {
+    function getEmitResolver(sourceFile: SourceFile, cancellationToken: CancellationToken, forceDtsEmit?: boolean) {
         // Ensure we have all the type information in place for this file so that all the
         // emitter questions of this resolver will return the right information.
-        getDiagnostics(sourceFile, cancellationToken);
+        if (!forceDtsEmit) getDiagnostics(sourceFile, cancellationToken);

```

Comment:
10/10 - last night I realized that having the check be dynamic with the `EmitResolver` but fed into `getDiagnostics` meant that we could cache the dummy `NodeFlags.TypeChecked` flag for files we'd force-skipped diagnostics for, but otherwise still need to check later, and unless you discarded the checker, that wasn't un-set-able. By checking this flag up here instead, we avoid setting the `TypeChecked` flag on a `SourceFile` we may still want to actually get diagnostics for later (but still skip the diagnostics walk, if not caching that skip), which is much better.